### PR TITLE
Npm workspaces

### LINKED ISSUE #346: Use NPM workspaces to manage multiple packages from within a singular top-level root package
https://github.blog/2020-10-13-presenting-v7-0-0-of-the-npm-cli/

Worth investigating ...

### PR DESCRIPTION
**Pull request by relu91**
This PR refactors the current project structure using npm workspaces. As a result of this refactoring `lerna` is no longer among our dev dependencies and everything has been ported to the new workspaces APIs. 

Open topics:
- Didn't test the publication procedure
- Didn't test CI (we'll see it after the creation of this PR)

Other points:
- After the refactoring npm added additional info to package.json of the bindings. I left that changes cause the were reasonable
- Opc-ua tests were not working, and I don't understand how the ever worked cause `url.origin` returns `'null'` with the current setup, and even testing online it returns the same ([see on runkit](https://runkit.com/embed/7grj9q5lpmxt)). I think the current code is more stable.

Closes #346

### CODE REVIEW

**Comment by danielpeintner** [comment]
@relu91 I think the [issues](https://github.com/eclipse/thingweb.node-wot/pull/483/checks?check_run_id=3472929027) appear for nodejs version 10
see https://github.com/eclipse/thingweb.node-wot/blob/master/.github/workflows/ci.yaml#L13

Shall we remove `10.x` and add `16.x` ?

**Comment by danielpeintner** [comment]
Note: I limited GitHub actions to LTS and current, see https://github.com/eclipse/thingweb.node-wot/commit/f398f79a3f567f2eebd7f260b25735a1cd896543

**Comment by egekorkan** [comment]
Sadly not a super useful comment but when we were restructuring TD playground, we had considered both options. I don't exactly remember why but lerna was better _at the time_. There is this SO thread from 9 months ago: https://stackoverflow.com/questions/64909635/is-lerna-needed-anymore-with-npm-7-0-0s-workspaces

I have also read that lerna is having difficulties maintaining itself but was there a reason for the change?

**Comment by relu91** [comment]
> @relu91 I think the issues appear for nodejs version 10
see https://github.com/eclipse/thingweb.node-wot/blob/master/.github/workflows/ci.yaml#L13

It might be related but I am not sure is completely the problem. Before lunch I was digging into it, I've just finished a call and I'll try to understand what's going on. 

> Shall we remove 10.x and add 16.x ?

Looking around 10.x is considered outdated. I would completely ok to remove the support for it in 0.8.x. Is everybody ok with it? Should we open an issue and call for opinions? 

>Sadly not a super useful comment but when we were restructuring TD playground, we had considered both options. I don't exactly remember why but lerna was better at the time. There is this SO thread from 9 months ago: https://stackoverflow.com/questions/64909635/is-lerna-needed-anymore-with-npm-7-0-0s-workspaces
>
>I have also read that lerna is having difficulties maintaining itself but was there a reason for the change?

It seems that we read the same stackoverflow question and took different decisions. What I read there is that `lerna` is still useful but for complementary tooling and commands. Tools and commands that we don't use. In `node-wot` we mostly used `lerna` as a device for handling our monorepo. Consequently, I think moving to `workspaces` is the right choice here. A couple of thoughts:
- It is better supported (I don't think `npm` is going to die soon, plus is well founded)
- I like the symlink approach (also used by yarn) -> less disk space used
- just one `pakcage-lock.json` 
- It should be well integrated with the publishing workflow

What do you think?

**Comment by egekorkan** [comment]
I see the advantages @relu91. We can just go ahead :)

**Comment by relu91** [comment]
Ok, I found out it's a well-known bug: https://github.com/npm/cli/issues/2632 . There's also a possible [workaround described](https://github.com/npm/cli/issues/2632#issuecomment-888960315), I am not a fan of it... but it is better than nothing 😄 .

**Comment by danielpeintner** [comment]
@relu91 merging in master causes merge conflicts.. sorry for that

**Comment by relu91** [comment]
don't worry not a big deal 😄

**Comment by danielpeintner** [comment]
Do you know why *no* GitHub jobs have been started?

Will do a local check also...

**Comment by relu91** [comment]
> Do you know why _no_ GitHub jobs have been started?
> 
> Will do a local check also...

It is probably because it was a merge commit... 🤔 but I checked on my fork. There was still a problem. Let's see if now it is solved

**Comment by relu91** [comment]
Just to keep you posted, one issue down. The current one is about typescript. It seems that in a clean repository it cannot find the types cause it actually builds the packages following the lexicographic order, not the dependency tree.

**Comment by relu91** [comment]
Ok, now it seems that the CI windows version does not like some native packages. I wasn't able to reproduce inside my machine yet. Let's see.

File: packages/binding-mbus/package.json [hunk 1]
```diff
@@ -35,5 +35,12 @@
   "scripts": {
     "build": "tsc",
     "test": "mocha --require ts-node/register test/*"
+  },
+  "bugs": {
+    "url": "https://github.com/thingweb/node-wot/issues"
+  },
+  "homepage": "https://github.com/thingweb/node-wot/tree/master/packages/binding-mbus#readme",
```
Comment (danielpeintner) [inline packages/binding-mbus/package.json:42]: Link is broken,  should be
https://github.com/eclipse/thingweb.node-wot/tree/master/packages/binding-mbus#readme

File: packages/binding-modbus/package.json [hunk 2]
```diff
@@ -38,5 +38,12 @@
   "scripts": {
     "build": "tsc",
     "test": "mocha --require ts-node/register test/*"
+  },
+  "bugs": {
+    "url": "https://github.com/thingweb/node-wot/issues"
+  },
+  "homepage": "https://github.com/thingweb/node-wot/tree/master/packages/binding-modbus#readme",
```
Comment (danielpeintner) [inline packages/binding-modbus/package.json:45]: Link broken also.. eclipse etc missing

File: packages/binding-netconf/package.json [hunk 3]
```diff
@@ -37,5 +37,13 @@
   "scripts": {
     "build": "tsc",
     "test": "mocha --require ts-node/register test/netconf-client-test.ts"
-  }
+  },
+  "bugs": {
+    "url": "https://github.com/thingweb/node-wot/issues"
+  },
+  "homepage": "https://github.com/thingweb/node-wot/tree/master/packages/binding-netconf#readme",
```
Comment (danielpeintner) [inline packages/binding-netconf/package.json:44]: Broken link

File: packages/binding-wotfirestore-browser-bundle/package.json [hunk 4]
```diff
@@ -2,7 +2,10 @@
   "name": "@node-wot/binding-wotfirestore-browser-bundle",
   "version": "0.8.0",
   "description": "A binding-wotfirestore bundle that can run in a web browser",
-  "repository": "https://github.com/hidetak/thingweb.node-wot",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/hidetak/thingweb.node-wot.git"
+  },
```
Comment (danielpeintner) [inline packages/binding-wotfirestore-browser-bundle/package.json:8]: Why does repository look different here, we seem to use something like this elsewhere

"repository": "https://github.com/eclipse/thingweb.node-wot/tree/master/packages/binding-wotfirestore",

File: packages/binding-wotfirestore-browser-bundle/package.json [hunk 5]
```diff
@@ -21,5 +24,10 @@
   },
   "scripts": {
     "build": "browserify -r vm:vm2 index.js --external coffee-script -o dist/binding-wotfirestore-bundle.js"
-  }
+  },
+  "bugs": {
+    "url": "https://github.com/hidetak/thingweb.node-wot/issues"
+  },
+  "homepage": "https://github.com/hidetak/thingweb.node-wot#readme",
+  "keywords": []
```
Comment (danielpeintner) [inline packages/binding-wotfirestore-browser-bundle/package.json:32]: Similar question, shouldn't it point to

https://github.com/eclipse/thingweb.node-wot/tree/master/packages/binding-wotfirestore

File: packages/binding-wotfirestore/package.json [hunk 6]
```diff
@@ -2,7 +2,10 @@
   "name": "@node-wot/binding-wotfirestore",
   "version": "0.8.0",
   "description": "Firestore binding for node-wot",
-  "repository": "https://github.com/hidetak/thingweb.node-wot",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/hidetak/thingweb.node-wot.git"
+  },
```
Comment (danielpeintner) [inline packages/binding-wotfirestore/package.json:8]: Similar comment, not pointing to package

File: packages/binding-wotfirestore/package.json [hunk 7]
```diff
@@ -38,5 +41,13 @@
     "typescript": "4.1.3",
     "typescript-standard": "0.3.36",
     "wot-typescript-definitions": "^0.8.0-SNAPSHOT.18"
-  }
+  },
+  "bugs": {
+    "url": "https://github.com/hidetak/thingweb.node-wot/issues"
+  },
+  "homepage": "https://github.com/hidetak/thingweb.node-wot#readme",
```
Comment (danielpeintner) [inline packages/binding-wotfirestore/package.json:48]: Similar comment, not pointing to package

**Review by danielpeintner — CHANGES_REQUESTED** [review]
See some broken links issues.
Also, I will run a local test and apart from that it looks good to me. thanks!

**Comment by danielpeintner** [comment]
> I'll try to fix the m-bus issues before the end of next week. I'll let you know when I'm done with this and with #475

Thanks @LennyHEVS 
I created https://github.com/eclipse/thingweb.node-wot/issues/488 to keep track

File: packages/binding-mbus/package.json [hunk 8]
```diff
@@ -35,5 +35,12 @@
   "scripts": {
     "build": "tsc",
     "test": "mocha --require ts-node/register test/*"
+  },
+  "bugs": {
+    "url": "https://github.com/thingweb/node-wot/issues"
```
Comment (danielpeintner) [inline packages/binding-mbus/package.json:40]: still wrong link

File: packages/binding-modbus/package.json [hunk 9]
```diff
@@ -38,5 +38,12 @@
   "scripts": {
     "build": "tsc",
     "test": "mocha --require ts-node/register test/*"
+  },
+  "bugs": {
+    "url": "https://github.com/thingweb/node-wot/issues"
```
Comment (danielpeintner) [inline packages/binding-modbus/package.json:43]: link is wrong

File: packages/binding-netconf/package.json [hunk 10]
```diff
@@ -37,5 +37,13 @@
   "scripts": {
     "build": "tsc",
     "test": "mocha --require ts-node/register test/netconf-client-test.ts"
-  }
+  },
+  "bugs": {
+    "url": "https://github.com/thingweb/node-wot/issues"
```
Comment (danielpeintner) [inline packages/binding-netconf/package.json:42]: wrong also

File: packages/binding-opcua/package.json [hunk 11]
```diff
@@ -39,5 +39,13 @@
   "scripts": {
     "build": "tsc",
     "test": "mocha --require ts-node/register test/opcua-client-test.ts"
-  }
+  },
+  "bugs": {
+    "url": "https://github.com/thingweb/node-wot/issues"
```
Comment (danielpeintner) [inline packages/binding-opcua/package.json:44]: to be fixed

**Comment by relu91** [comment]
Another update, I tried to re-introduce mbus cause I think the real problem is the fact that the ci does not have the right window-build-tools installed. However, it seems that installing those tools on the ci is causing an infinite loop... I'll revert the commit and address @danielpeintner comments above.

**Comment by danielpeintner** [comment]
I think we can go ahead merging this PR "as is" and solve mbus issues in a separate PR.
@relu91 if you agree please merge 👍

**Pull request merged**